libs/vkd3d/cache.c

```diff
@@ -302,10 +307,29 @@ static HRESULT STDMETHODCALLTYPE d3d12_pipeline_library_SetPrivateData(d3d12_pip
         REFGUID guid, UINT data_size, const void *data)
 {
     struct d3d12_pipeline_library *pipeline_library = impl_from_ID3D12PipelineLibrary(iface);
+    const char *name;
+    HRESULT hr;
 
     TRACE("iface %p, guid %s, data_size %u, data %p.\n", iface, debugstr_guid(guid), data_size, data);
 
-    return vkd3d_set_private_data(&pipeline_library->private_store, guid, data_size, data);
+    if (FAILED(hr = vkd3d_private_data_lock(&pipeline_library->private_store)))
+        return hr;
+
+    if (FAILED(hr = vkd3d_set_private_data(&pipeline_library->private_store, guid, data_size, data)))
+    {
+        vkd3d_private_data_unlock(&pipeline_library->private_store);
+        return hr;
+    }
+
+    if (vkd3d_private_data_object_name_ptr(guid, data_size, data, &name))
```

Comment:
I'm not a fan of the intense code duplication here. I know callbacks came up in discussions, and I think I would favor callbacks or some interface. If this were C++, I would have done something like:

SetPrivateData is implemented in ID3D12Object, then that calls a private virtual IVKD3DNameListener::OnObjectName or something like that when we observe the GUID in question, but setting all that machinery up in C is a ton of boilerplate as well ...

Callbacks would avoid most of that boilerplate, and I think I would favor that. It's consumed immediately in vkd3d_set_private_data after all. It's not worse than qsort in that sense.

Reply:
Either that, or something like:

```c
static HRESULT STDMETHODCALLTYPE d3d12_device_SetPrivateData(d3d12_device_iface *iface,
        REFGUID guid, UINT data_size, const void *data)
{
    struct d3d12_device *device = impl_from_ID3D12Device(iface);

    TRACE("iface %p, guid %s, data %p.\n", iface, debugstr_guid(guid), data);

    return vkd3d_set_private_data(&device->private_store, guid, data_size, data, d3d12_device_set_name);
}
```

I don't think we need another interface for that, we can just use the existing stuff and add an argument.

Reply:
Yes, just need to add the `device` argument so the d3d12_device_set_name gets the pointer it expects, but that's probably the sanest solution we can do in C.

Reply:
Resolved locally.